MINOR: System tests for Raft-based metadata quorums

rondagostino opened the pull request:
Adds a new `sanity_checks/bounce_test.py` system test for a simple produce/bounce/produce series of events.  Augments this and other `sanity_checks` tests so they run all metadata quorum types: ZooKeeper, remote Raft, and co-located Raft.

Augments several `tests/client` and `tests/core` system tests to run for remote Raft-based metadata quorums in addition to ZooKeeper.  Co-located Raft controllers are not tested here as this configuration is presumed to work assuming `sanity_checks` as well as any unit/integration tests pass.

Augments a simple set of tests in `tests/connect`, `tests/streams`, and `tests/tools` to run for remote Raft-based metadata quorums in addition to ZooKeeper.

Adds missing `@cluster` annotations for dozens of system tests that were missing them -- this was causing these tests to grab the entire cluster of nodes and negatively impact the parallelism of the system test run.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

cmccabe committed:
KAFKA-12276: Add the KIP-500 controller code

cmccabe committed:
KAFKA-12276: Add the KIP-500 controller code

cmccabe committed:
KAFKA-12276: Add the KIP-500 controller code

cmccabe committed:
KAFKA-12276: Add the KIP-500 controller code

rondagostino commented:
There are 6 failing tests across these 3 test classes:

**kafka.server.KafkaConfigTest**
```
KafkaConfigTest > testFromPropsInvalid() FAILED
    org.opentest4j.AssertionFailedError: Expected exception for property `broker.heartbeat.interval.ms` with invalid value `-1` was not thrown ==> Expected java.lang.Exception to be thrown, but nothing was thrown.
```

**AutoTopicCreationManagerTest**
```
AutoTopicCreationManagerTest > testCreateTxnTopic() FAILED
    Wanted but not invoked:
    brokerToControllerChannelManager.sendRequest(
        CreateTopicsRequestData(topics=[CreatableTopic(name='__transaction_state', numPartitions=2, replicationFactor=2, assignments=[], configs=[])], timeoutMs=100, validateOnly=false),
        <any kafka.server.ControllerRequestCompletionHandler>
    );
    -> at kafka.server.AutoTopicCreationManagerTest.testCreateTopic(AutoTopicCreationManagerTest.scala:119)
    Actually, there were zero interactions with this mock.

AutoTopicCreationManagerTest > testCreateOffsetTopic() FAILED
    Wanted but not invoked:
    brokerToControllerChannelManager.sendRequest(
        CreateTopicsRequestData(topics=[CreatableTopic(name='__consumer_offsets', numPartitions=2, replicationFactor=2, assignments=[], configs=[])], timeoutMs=100, validateOnly=false),
        <any kafka.server.ControllerRequestCompletionHandler>
    );
    -> at kafka.server.AutoTopicCreationManagerTest.testCreateTopic(AutoTopicCreationManagerTest.scala:119)
    Actually, there were zero interactions with this mock.

AutoTopicCreationManagerTest > testCreateNonInternalTopic() FAILED
    Wanted but not invoked:
    brokerToControllerChannelManager.sendRequest(
        CreateTopicsRequestData(topics=[CreatableTopic(name='topic', numPartitions=1, replicationFactor=1, assignments=[], configs=[])], timeoutMs=100, validateOnly=false),
        <any kafka.server.ControllerRequestCompletionHandler>
    );
    -> at kafka.server.AutoTopicCreationManagerTest.testCreateTopic(AutoTopicCreationManagerTest.scala:119)
    Actually, there were zero interactions with this mock.
```

**MetadataRequestWithForwardingTest**
```
MetadataRequestWithForwardingTest > testAutoCreateOfCollidingTopics() FAILED
    org.opentest4j.AssertionFailedError: expected: <Set(LEADER_NOT_AVAILABLE, INVALID_TOPIC_EXCEPTION)> but was: <Set(UNKNOWN_TOPIC_OR_PARTITION)>

MetadataRequestWithForwardingTest > testAutoTopicCreation() FAILED
    org.opentest4j.AssertionFailedError: expected: <LEADER_NOT_AVAILABLE> but was: <UNKNOWN_TOPIC_OR_PARTITION>
```

rondagostino committed:
MINOR: Add KIP-500 Broker

No need to invoke dynamicConfig.initialize() at all when using Raft

Fix formatting and make variables private as per review

Revert optional nodeId in SocketServer: aliasing makes it unnecessary

Exclude Raft controller nodes from Produce magic version determination

Fix responsibility for startup/shutdown of alterisr/forwarding managers

Fix BrokerServer log ident

Invoke replicaManager.endMetadataChangeDeferral() before starting up

Fix compile error

Return random active broker ID for Raft-based controller ID

Ignore reserved.broker.max.id for Raft-based cases

Cleanup KafkaConfig and tests related to {broker,node}.id

Fix bug in initializeLogDirs()

Add NetworkChannel.updateEndpoint(int id, RaftConfig.InetAddressSpec address)

fix compile error

revert old forwarding code

Refactor envelope handling logic into a common place

Add EnvelopeUtils

Add ControllerServer

rondagostino committed:
Remove unused imports

rondagostino committed:
Instantiate QuorumController instance

rondagostino committed:
Add test_bounce.py system test

rondagostino committed:
Fix NPE due to creating AutoTopicCreateMgr before Grp/Txn Coordinators

rondagostino committed:
Workaround Raft issue

rondagostino committed:
Merge commit '8f6e443e02174cfcc4eb29dd383269f09ccfc3b7' into kip500_full2

rondagostino committed:
Set QuorumController metrics

rondagostino committed:
Replace configuredLogContext param with controllerSocketServer: Boolean

rondagostino committed:
Merge remote-tracking branch 'apache/trunk' into kip500_full

cmccabe committed:
Address review comments

rondagostino committed:
Fix Raft broker restart issue when offset partitions are deferred

cmccabe committed:
Address review comments

rondagostino committed:
System test Raft-based clusters: sanity_checks and tests/client

rondagostino committed:
Merge remote-tracking branch 'apache/trunk' into kip500_full

rondagostino committed:
Fix typo causing tests to not run

ijuma commented on build.gradle:
```diff
@@ -1029,6 +1029,7 @@ project(':metadata') {
     compile project(':clients')
     compile libs.jacksonDatabind
     compile libs.jacksonJDK8Datatypes
+    compile libs.metrics
```
Let's file a JIRA to discuss whether the metadata module should use Kafka metrics or Yammer metrics.

rondagostino replied:
> file a JIRA to discuss whether the metadata module should use Kafka metrics or Yammer metrics.

https://issues.apache.org/jira/browse/KAFKA-12348

cc: @cmccabe

cmccabe replied:
thanks. I think this will be a post-2.8 thing

rondagostino committed:
Add KIP-500 system tests to tests/core plus connect/streams smoke tests

cmccabe committed:
Some partition management improvements

Make epoch bumps implicit in PartitionChangeRecord

Always bump the leader epoch of a partition led by a broker in controlled shutdown

cmccabe committed:
Shut down the broker metadata listener during controlled shutdown

cmccabe committed:
fix leader election

rondagostino committed:
Merge commit '293841b05d881a83c7e80631a20bcac939534681' into kip500_full

rondagostino committed:
Merge commit 'd2a78075b71197c634f83188b4d6b07912d269bd' into kip500_full

rondagostino committed:
Merge commit '807b7cc2393eb5c1a291854bd7cb3e2df2f9d1dd' into kip500_full

rondagostino committed:
Merge commit '435695488f48c49babfa7f297a9123bd1c4a69ae' into kip500_full

rondagostino committed:
Merge commit 'ff2eef63ed1d78398da8ca456b978180e86bf23f' into kip500_full

rondagostino committed:
Merge commit '87148a92ab4d736e02dbc7382c2d19a669d7ae8e' into kip500_full

rondagostino committed:
Add missing @cluster annotations

rondagostino committed:
Add missing import

rondagostino committed:
Merge remote-tracking branch 'apache/trunk' into kip500_full

rondagostino committed:
Revert workaround now that KAFKA-12331 is merged

rondagostino commented:
http://confluent-kafka-branch-builder-system-test-results.s3-us-west-2.amazonaws.com/2021-02-20--001.1613812911--rondagostino--kip500_full--f59453faa/report.html

Tests|Passes|Failures|Ignored|Time
940|558|185|197|359 minutes

Some observations.  First, adding the missing @cluster annotations really helped.  Running almost 750 tests in under 6 hours is a dramatic improvement.  Here’s the stats from a run 2 days ago:

Tests|Passes|Failures|Ignored|Time
729|427|105|197|388 minutes

We are running 743/532 = 40% more tests, but we are able to finish in 10% less time.  So that’s a big win.

We have 130 more tests passing and 80 more tests failing.  It is VERY difficult to find the signal amid so much noise.  A quick browse through our results shows lots of test failures that are not augmented for Raft at all, and many test that are augmented are failing for both ZK and Raft cases.

cmccabe commented on tests/kafkatest/services/console_consumer.py:
```diff
@@ -151,7 +151,10 @@ def prop_file(self, node):
+        if self.new_consumer:
+            args['broker_list'] = self.kafka.bootstrap_servers(self.security_config.security_protocol)
+        else:
+            args['zk_connect'] = self.kafka.zk_connect_setting()
```
This is a slight behavior change for the old consumer, right?  Previously we set both `broker_list` and `zk_connect`, but now we set only `zk_connect`. Can we keep setting both for now? (I agree we should skip setting zk_connect for the new consumer)

cmccabe commented on tests/kafkatest/services/performance/end_to_end_latency.py:
```diff
@@ -74,12 +74,15 @@ def __init__(self, context, num_nodes, kafka, topic, num_records, compression_ty
             'kafka_run_class': self.path.script("kafka-run-class.sh", node),
             'java_class_name': self.java_class_name()
         })
+        if node.version < V_0_9_0_0:
```
It would be good to have a function in `version.py` to reflect this

rondagostino replied:
> would be good to have a function in version.py to reflect [functionality unsupported below 0.9.0]

I'll add this function:

```
    def supports_bootstrap_server(self):
        return self >= V_0_9_0_0
```

I then made the changes all over since this 0.9.0 constant is checked in several places:

```
tests/kafkatest/services/console_consumer.py
tests/kafkatest/services/performance/consumer_performance.py
tests/kafkatest/services/performance/end_to_end_latency.py
tests/kafkatest/services/performance/producer_performance.py
tests/kafkatest/tests/core/upgrade_test.py
tests/kafkatest/tests/core/upgrade_test.py
```

We can revert if you think I went too far.

cmccabe replied:
thanks.  that sounds good!

can you name it something like `consumer_supports_bootstrap_server`?  There are a bunch of other bootstrap server functions (`acl_command_supports_bootstrap_server`, `topic_command_supports_bootstrap_server`, etc.) so it would be good to be clear

rondagostino replied:
> name it something like consumer_supports_bootstrap_server... bunch of other bootstrap server functions

Good point.  I renamed it `consumer_supports_bootstrap_server()` as you suggested.

cmccabe commented on tests/kafkatest/tests/client/client_compatibility_features_test.py:
```diff
@@ -107,7 +108,9 @@ def invoke_compatibility_program(self, features):
-    @parametrize(broker_version=str(DEV_BRANCH))
+    @cluster(num_nodes=7)
+    @matrix(broker_version=[str(DEV_BRANCH)], metadata_quorum=quorum.all_non_upgrade)
+    @cluster(num_nodes=7)
```
It seems like `@cluster` is repeated twice here.  We only need it once, right?

cmccabe commented on tests/kafkatest/tests/client/client_compatibility_features_test.py:
```diff
@@ -120,10 +123,9 @@ def invoke_compatibility_program(self, features):
     @parametrize(broker_version=str(LATEST_2_4))
     @parametrize(broker_version=str(LATEST_2_5))
-    @parametrize(broker_version=str(LATEST_2_6))
-    @parametrize(broker_version=str(LATEST_2_7))
```
hmm.  we don't want to drop 2.6 and 2.7 here, do we?

cmccabe commented on tests/kafkatest/tests/client/client_compatibility_produce_consume_test.py:
```diff
@@ -46,13 +47,16 @@ def __init__(self, test_context):
-    @parametrize(broker_version=str(DEV_BRANCH))
+    @cluster(num_nodes=9)
+    @matrix(broker_version=[str(DEV_BRANCH)], metadata_quorum=quorum.all_non_upgrade)
+    @cluster(num_nodes=9)
```
`@cluster` is repeated again

cmccabe commented on tests/kafkatest/tests/core/round_trip_fault_test.py:
```diff
@@ -47,32 +55,48 @@ def __init__(self, test_context):
+        elif quorum.for_test(self.test_context) == quorum.remote_raft:
+            return self.kafka.controller_quorum.nodes
+        else: # co-located case, which we currently don't test but handle here for completeness in case we do test it
+            return []
```
should this throw an exception? Or does the current code actually work for this case?

rondagostino replied:
> throw an exception or does the current code actually work for this case?

The code always needs `<Kafka nodes>` + `<remote quorum nodes>`, where the latter are the ZooKeeper or remote Controller Quorum nodes.  If we were to run this test with co-located Raft Quorum controllers then those nodes would be accounted for because they are part of the Kafka nodes, so there is no need to explicitly add them.  So this code is correct in that it return an empty list for that case.  As was indicated in the comment, it's here just in case we ever decide we want to test it.

cmccabe replied:
sounds good

cmccabe commented:
Good work spotting the missing `@cluster` annotations! That is quite a performance win. I wonder if we should disallow tests without this annotation in the future.

rondagostino committed:
Set both broker_list and zk_connect for old consumer

rondagostino committed:
Add version.supports_bootstrap_server()

rondagostino commented:
> I wonder if we should disallow tests without `[the @cluster]` annotation in the future.

I think perhaps yes.  Is there a need to allow it?  A simple oversight generates a significant parallelism hit at this point -- and they compound quickly as evidenced here.

rondagostino committed:
minor client compatibility test fixups

rondagostino committed:
Rename version.supports_bootstrap_server => consumer_supports_bootstrap_server

cmccabe reviewed (approved):
LGTM. Thanks, @rondagostino !
